### PR TITLE
SSO missing token when trying to sign in 

### PR DESCRIPTION
**Pull request by nicocucuzza**
<ins>**Problem**</ins>

- When using SSO, users sometimes see the message `The token field is required when token is not present`
- Plugin was relying on the response of the auth token request. If there was an error, it tries to open the domain using an empty token.

<ins>**Solution**</ins>

- Improve the error handling after requesting the auth-token

### CODE REVIEW

**Comment by nicocucuzza** [comment]
[(4) SSO missing token when trying to sign in [4]](https://trello.com/c/iRMvaJo6/4060-4-sso-missing-token-when-trying-to-sign-in-4)

File: public/class-award-force-api-v2.php [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 class AwardForceAPIV2 {
 
-    private $apiUrl = 'https://api.awardsplatform.com';
+    private $apiUrl = 'https://api.staging.awardsplatform.com';
```
Comment (eyalgrossdev) [inline public/class-award-force-api-v2.php]: Why was this changed to staging? Just for testing?

File: public/class-award-force-sso.php [hunk 2]
```diff
@@ -84,17 +93,35 @@ private function requestSlugByEmail($email)
...
+                }
+                $this->api->handleException(new Exception($response->message));
             }
-
             sleep(1);
             $retries--;
         }
+
+        if (!$token) {
+            $this->api->handleException(new Exception('There was an error requesting a token to Award Force'));
```
Comment (eyalgrossdev) [inline public/class-award-force-sso.php]: _There was an error requesting a token to Award Force_

should be

_There was an error requesting a token from Award Force_

File: public/class-award-force-sso.php [hunk 3]
```diff
@@ -70,7 +75,11 @@ private function requestSlug(WP_User $user)
             'password' => uniqid(),
         ]);
 
-        return $response->slug;
+        if (isset($response->slug)) {
```
Comment (eyalgrossdev) [inline public/class-award-force-sso.php]: I'd check for and handle the exception here and return the slug as the happy path.

```
if (!isset($response->slug)) {
   $this->api->handleException(new Exception($response->message ?: 'There was an error creating the user.'));
}

return $response->slug;
```
Also above I would put the create user api call in it's own function and call that returning the response for evaluation here.

File: public/class-award-force-sso.php [hunk 4]
```diff
@@ -84,17 +93,35 @@ private function requestSlugByEmail($email)
...
+
         while ($retries > 0) {
-            $response = $this->api->get('/user/' . $slug . '/auth-token');
-            if ($token = $response->auth_token) {
-                return $token;
+            if ($response = $this->sendAuthTokenRequest($slug)) {
+                if ($token = $response->auth_token) {
+                    return $token;
+                }
+                $this->api->handleException(new Exception($response->message));
```
Comment (eyalgrossdev) [inline public/class-award-force-sso.php:111]: An exception will break out of the retry loop? Is that what you want here?

**Review by eyalgrossdev — APPROVED** [review]
Looks good @nicocucuzza . Some comments / questions but nothing major. I'll leave this to Rachel to QA.

File: public/class-award-force-sso.php [hunk 5]
```diff
@@ -59,7 +60,11 @@ private function requestSlug(WP_User $user)
     {
         $response = $this->requestSlugByEmail($user->user_email);
 
-        if ($response->slug) {
+        if (isset($response->status_code) && $response->status_code != 200) {
```
Comment (eyalgrossdev) [inline public/class-award-force-sso.php]: @nicocucuzza as per the thread in slack... I think these checks here will mean it will never get passed this point and create the user if it doesn't exist.

File: public/class-award-force-sso.php (hunk 4 above)
Comment (nicocucuzza) [reply public/class-award-force-sso.php:111]: Yes, retry only if we don't obtain a response. In case of error, log the message and abort

File: public/class-award-force-sso.php (hunk 3 above)
Comment (nicocucuzza) [reply public/class-award-force-sso.php]: :+1:

File: public/class-award-force-sso.php (hunk 2 above)
Comment (nicocucuzza) [reply public/class-award-force-sso.php]: :+1:
